### ISSUE TITLE
Equals parsing incompatibility

### ISSUE THREAD

**Issue by RobSykes** [oilshell/oil#963]
BASH_VERSION='5.0.17(1)-release':

$ =() { echo it works;}
$ =
it works

OIL_VERSION=0.8.12 (osh):

osh$ =() { echo it works;}
  =() { echo it works;}
    ^
[ interactive ]:2: Syntax error in expression (near Id.Op_RParen)

**Comment by andychu** [comment]
This is intentional, although the error message should be better.

The goal of Oil is to run real shell programs programs ([and it does](https://www.oilshell.org/blog/2018/01/15.html)), not be compatible with any bash snippet that can be synthesized.   This is a big difference!

I somewhat documented this under "Common Subset Principle": https://github.com/oilshell/oil/wiki/Language-Design-Principles

Sometimes you have to make tiny changes to a real shell program.  But if it still runs under both bash and OSH, and is readable/natural, that's working as intended.

It's also somewhat documented here, since `=` is a keyword that prints an expression, e.g. try `= 1 + 2 * 3` (sort of like Lua). 

https://www.oilshell.org/release/0.8.pre1/doc/known-differences.html#oil-has-more-keywords-which-shadow-builtins-functions-and-commands

So basically it will help to mention context, and incompatibilities in real programs is prioritized.  Thanks again for trying it!

**Comment by andychu** [comment]
I will close this for now, but please let me know if you have any other issues with Oil / questions about it

**Comment by RobSykes** [comment]
The background is: I was following the "Gradually Upgrading Shell to Oil" which stated "Your shell script should run as is. You most likely won't notice a difference, except perhaps better error messages when the script fails.".  The first script I tried was my standard .bashrc (which naturally, doesn’t fail under bash, and which defines a function called '='—not the definition in the OP by the way, which was simplified for brevity): it failed under osh—given the statements above, this was unexpected.

Having now looked at the "Known Differences Between OSH and Other Shells", it’s statement "OSH is meant to run all POSIX shell programs, and most bash programs." is much clearer—perhaps this could be reiterated in the Upgrading Shell document.

**Comment by andychu** [comment]
Ah OK thanks for the context.  I updated the wiki page.

If there were a very important script or many people using `=()`, it might be worth changing, but I think it's pretty rare, and shells like dash and zsh don't support it.
